lxd/main_callhook.go

```diff
@@ -24,14 +32,175 @@ func (c *cmdCallhook) Command() *cobra.Command {
   Call container lifecycle hook in LXD
 
   This internal command notifies LXD about a container lifecycle event
-  (start, stopns, stop, restart) and blocks until LXD has processed it.
+  (start, startmountns, stopns, stop, restart) and blocks until LXD has processed it.
 `
 	cmd.RunE = c.Run
 	cmd.Hidden = true
 
+	// devicesRootFolder is used to specify where to look for CDI config device files.
+	cmd.Flags().StringVar(&c.devicesRootFolder, "devicesRootFolder", "", "Root folder for CDI devices")
+
 	return cmd
 }
 
+// resolveTargetRelativeToLink converts a target relative to a link path into an absolute path.
+func resolveTargetRelativeToLink(link string, target string) (string, error) {
+	if !filepath.IsAbs(link) {
+		return "", fmt.Errorf("The link must be an absolute path: %q (target: %q)", link, target)
+	}
+
+	if filepath.IsAbs(target) {
+		return target, nil
+	}
+
+	linkDir := filepath.Dir(link)
+	absTarget := filepath.Join(linkDir, target)
+	cleanPath := filepath.Clean(absTarget)
+	absPath, err := filepath.Abs(cleanPath)
+	if err != nil {
+		return "", err
+	}
+
+	return absPath, nil
+}
+
+// customCDILinkerConfFile is the name of the linker conf file we will write to
+// inside the container. The `00-lxdcdi` prefix is chosen to ensure that these libraries have
+// a higher precedence than other libraries on the system.
+var customCDILinkerConfFile = "00-lxdcdi.conf"
+
+// applyCDIHooksToContainer is called before the container has started but after the container namespace has been setup,
+// and is used whenever CDI devices are added to a container and where symlinks and linker cache entries need to be created.
+// These entries are listed in a 'CDI hooks file' located at `hooksFilePath`.
+func applyCDIHooksToContainer(devicesRootFolder string, hooksFilePath string) error {
+	hookFile, err := os.Open(filepath.Join(devicesRootFolder, hooksFilePath))
+	if err != nil {
+		return fmt.Errorf("Failed to open the CDI hooks file at %s: %w\n", hooksFilePath, err)
```

Comment:
why do the errors in this function add a `\n` to the error?

Reply:
no reason :) fixing that.

Reply:
fixed